File: src/test/java/nl/iljabooij/garmintrainer/model/SpeedTest.java
/*
 * Copyright 2009 Ilja Booij
 * 
 * This file is part of GarminTrainer.
 * 
 * GarminTrainer is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * GarminTrainer is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with GarminTrainer.  If not, see <http://www.gnu.org/licenses/>.
 */
package nl.iljabooij.garmintrainer.model;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import nl.iljabooij.garmintrainer.model.Speed.Unit;

import org.joda.time.Duration;
import org.junit.Test;


public class SpeedTest {
	private static final double DELTA = 0.00001;
	
	private static final Length _100Meters = Length.createLengthInMeters(100);
	private static final Duration _10Seconds = Duration.standardSeconds(10);
	private static final Length _36Km = Length.createLength(36, Length.Unit.Kilometer);
	private static final Duration _1Hour = Duration.standardHours(1);
	
	@Test
	public void createMetersPerSecond() {
		final double value = 21.0;
		
		final Speed speed = Speed.createExactSpeedInMetersPerSecond(value);
		assertEquals("correct number of meters per second", value,
				speed.getValueInMetersPerSecond(), DELTA);
	}
	
	@Test
	public void createSpeedInMetersPerSecond() {
		final Speed speed10MPS = Speed.createSpeedInMetersPerSecond(_100Meters, _10Seconds);
		final Speed _36KmH = Speed.createSpeedInMetersPerSecond(_36Km, _1Hour);
		assertEquals("100 meters in 10 seconds equals 10 m/s", 10.0,
				speed10MPS.getValueInMetersPerSecond(), DELTA);
		assertEquals("36 km in 1 hour equals 10 m/s", 10.0,
				_36KmH.getValueInMetersPerSecond(), DELTA);
	}
	
	@Test
	public void createExactSpeed() {
		final Speed _10MPS = Speed.createExactSpeed(10.0, Unit.MetersPerSecond);
		final Speed _36KmH = Speed.createExactSpeed(36.0, Unit.KilometersPerHour);
		
		assertEquals(_10MPS, _36KmH);
	}
	
	@Test
	public void createSpeed() {
		final Speed _10MPS = Speed.createSpeed(_36Km, _1Hour, Unit.MetersPerSecond);
		assertEquals("100 meters in 10 seconds equals 10 m/s", 10.0,
				_10MPS.getValueInMetersPerSecond(), DELTA);
		assertEquals("100 meters in 10 seconds equals 10 m/s", 10.0,
				_10MPS.getValue(), DELTA);
		assertEquals("unit is Unit.MetersPerSecond", Unit.MetersPerSecond, 
				_10MPS.getUnit());
		
		final Speed _36KmH = Speed.createSpeed(_36Km, _1Hour, Unit.KilometersPerHour);
		assertEquals("36 km/h equals 10 m/s", 10.0,
				_36KmH.getValueInMetersPerSecond(), DELTA);
		assertEquals("should read 36 in km/h", 36,
				_36KmH.getValue(), DELTA);
		assertEquals("check that the right unit is returned", Unit.KilometersPerHour, 
				_36KmH.getUnit());
	}
	
	@Test(expected=NullPointerException.class)
	public void createSpeedThrowsNPEOnNullUnit() {
		Speed.createSpeed(_36Km, _1Hour, null);
	}
	
	@Test(expected=NullPointerException.class)
	public void createSpeedWithNullLengthShouldThrowNPE() {
		Speed.createSpeedInMetersPerSecond(null, _10Seconds);
	}
	
	@Test(expected=NullPointerException.class)
	public void createSpeedWithNullDurationShouldThrowNPE() {
		Speed.createSpeedInMetersPerSecond(_100Meters, null);
	}
	
	@Test
	public void getValue() {
		final Speed speed10MPS = Speed.createSpeedInMetersPerSecond(_100Meters, _10Seconds);
		
		final double value = speed10MPS.getValue(Unit.KilometersPerHour);
		assertEquals("10 m/s is 36 km/h", 36.00,
				value, DELTA);
	}
	
	@Test(expected=NullPointerException.class) 
	public void getValueWithNullUnitThrowsNPE() {
		Speed.createSpeedInMetersPerSecond(_100Meters, _10Seconds).getValue(null);
	}
	
	@Test
	public void equalsWorks() {
		final Speed _10MPS = Speed.createSpeedInMetersPerSecond(_100Meters, _10Seconds);
		final Speed _36KmH = Speed.createSpeedInMetersPerSecond(_36Km, _1Hour);
		
		assertEquals("speeds should be equal", _10MPS, _36KmH);
	}
	
	@Test
	public void hashCodeWorks() {
		final Speed _10MPS = Speed.createSpeedInMetersPerSecond(_100Meters, _10Seconds);
		final Speed _36KmH = Speed.createSpeedInMetersPerSecond(_36Km, _1Hour);
		
		assertEquals("hashcode of speeds should be equal", _10MPS.hashCode(), _36KmH.hashCode());
	}
	
	@Test
	public void toStringWorks() {
		final Speed _10MPS = Speed.createSpeedInMetersPerSecond(_100Meters, _10Seconds);
		final Speed _36KmH = Speed.createSpeed(_36Km, _1Hour, Unit.KilometersPerHour);
		
		assertEquals("speed string ok", "10.0 m/s", _10MPS.toString());
		assertEquals("speed string ok", "36.00 km/h", _36KmH.toString());
	}
	
	@Test
	public void convertWorks() {
		final Speed _10MPS = Speed.createSpeedInMetersPerSecond(_100Meters, _10Seconds);
		final Speed _36KmH = _10MPS.convert(Unit.KilometersPerHour);
		
		assertEquals("speed string ok", "36.00 km/h", _36KmH.toString());
	}
	
	@Test(expected=NullPointerException.class)
	public void convertThrowsNPEOnNullUnit() {
		final Speed _10MPS = Speed.createSpeedInMetersPerSecond(_100Meters, _10Seconds);
		_10MPS.convert(null);
	}
	
	@Test
	public void testCompareTo() {
		final Speed _10MPS = Speed.createSpeedInMetersPerSecond(_100Meters, _10Seconds);
		final Speed _36KmH = _10MPS.convert(Unit.KilometersPerHour);
		final Speed _11MPS = Speed.createSpeedInMetersPerSecond(Length.createLengthInMeters(110), _10Seconds);
		final Speed _12MPS = Speed.createSpeedInMetersPerSecond(Length.createLengthInMeters(120), _10Seconds);
		
		assertTrue("10 m/s == 36 km/h", _10MPS.compareTo(_36KmH) == 0);
		assertTrue("10 m/s < 11 m/s",_10MPS.compareTo(_11MPS) < 0);
		assertTrue("11 m/s < 12 m/s",_11MPS.compareTo(_12MPS) < 0);
		assertTrue("12 m/s > 10 m/s",_12MPS.compareTo(_10MPS) > 0);
	}
	
	@Test(expected=NullPointerException.class)
	public void compareToThrowsNPEOnNullArgument() {
		final Speed _10MPS = Speed.createSpeedInMetersPerSecond(_100Meters, _10Seconds);
		_10MPS.compareTo(null);
	}
}

File: src/test/scala/nl/iljabooij/garmintrainer/model/SpeedTest.scala
/*
 * Copyright 2009 Ilja Booij
 * 
 * This file is part of GarminTrainer.
 * 
 * GarminTrainer is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * GarminTrainer is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with GarminTrainer.  If not, see <http://www.gnu.org/licenses/>.
 */
package nl.iljabooij.garmintrainer.model

import nl.iljabooij.garmintrainer.model.Duration._
import nl.iljabooij.garmintrainer.model.Length.{Kilometer,Meter}
import nl.iljabooij.garmintrainer.model.Speed.{KilometersPerHour,MetersPerSecond}

import org.junit.Assert._
import org.mockito.Mockito._
import org.scalatest.junit.{JUnit3Suite,AssertionsForJUnit}
import org.scalatest.mock.MockitoSugar


class SpeedTest extends JUnit3Suite with AssertionsForJUnit with MockitoSugar {
  val DELTA = 0.00001
  
  def testSiValues {
    val speedInKmph = 30.22
    val kmph = new KilometersPerHour(speedInKmph)
    assertEquals(speedInKmph/3.6, kmph.siValue, DELTA)
    
    val speedInMps = 10.3
    assertEquals(speedInMps, new MetersPerSecond(speedInMps).siValue, DELTA)
  }
  
  def testCalculateSpeed {
    val speed = Speed.speed(new Meter(10), second)
    assertEquals(new MetersPerSecond(10), speed)
    assertEquals(new MetersPerSecond(10), Speed.
    	speed(new Kilometer(36), hour))
  }
  
  def testHashCode {
    assertEquals(new MetersPerSecond(10).hashCode, Speed.
    	speed(new Kilometer(36), hour).hashCode)
  }
  
  def testToMetersPerSecond {
    val kmph = new KilometersPerHour(36)
    val mps = kmph.toMetersPerSecond
    assertEquals(10.0, mps.mps, DELTA)
  }
  
  def testToKilometersPerHour {
    val mps = new MetersPerSecond(10)
    val kmph = mps.toKilometersPerHour
    assertEquals(36.0, kmph.kmph, DELTA)
  }  
  
  def testToString {
    assertEquals("10.0 m/s", new MetersPerSecond(10.0).toString)
    assertEquals("-10.0 m/s", new MetersPerSecond(-10.0).toString)
    assertEquals("0.2 m/s", new MetersPerSecond(0.23).toString)
    assertEquals("10.0 km/h", new KilometersPerHour(10.0).toString)
    assertEquals("-10.0 km/h", new KilometersPerHour(-10.0).toString)
    assertEquals("0.2 km/h", new KilometersPerHour(0.23).toString)
  }
  
  def testSmallerThan {
    val _10_1mps = new MetersPerSecond(10.1)
    val _36kmph = new KilometersPerHour(36.0)
    assertFalse(_10_1mps < _36kmph)
    assertTrue(_36kmph < _10_1mps)	
  }
  
  def testGreaterThan {
    val _10_1mps = new MetersPerSecond(10.1)
    val _36kmph = new KilometersPerHour(36.0)
    assertTrue(_10_1mps > _36kmph)
    assertFalse(_36kmph > _10_1mps)	
  }
  
  def testMax {
    val _10_1mps = new MetersPerSecond(10.1)
    val _36kmph = new KilometersPerHour(36.0)
    assertEquals(_10_1mps, _10_1mps max _36kmph)
    assertEquals(_10_1mps, _36kmph max _10_1mps)
  }
  
  def testMin {
    val _10_1mps = new MetersPerSecond(10.1)
    val _36kmph = new KilometersPerHour(36.0)
    assertEquals(_36kmph, _10_1mps min _36kmph)
    assertEquals(_36kmph, _36kmph min _10_1mps)
  }
}
